docs/caching.md

```diff
@@ -0,0 +1,281 @@
+    time.sleep(2)  # This makes the function take 2s to run
+    return inner_func(a, b) + 1
+
+a = 2
+b = 21
+res = expensive_computation(a, b)
+
+st.write("Result:", res)
+```
+
+Even though `inner_func()` is not annotated with [`@st.cache`](api.html#streamlit.cache), when we edit its body we cause a "Cache miss" in the outer `expensive_computation()`.
+
+That's because Streamlit always traverses your code and its dependencies to verify that the cached values are still valid. This means that while developing your app you can edit your code freely without worrying about the cache. Any change you make to your app, Streamlit should do the right thing!
+
+Streamlit is also smart enough to only traverse dependencies that belong to your app, and skip over any dependency that comes from an installed Python library.
+
+
+## Example 5: Use the global cache to speed up your app for all users
+
+Going back to our original function, let's add a widget to control the value of `b`:
+
+```python
+import streamlit as st
+import time
+
+@st.cache(suppress_st_warning=True)
+def expensive_computation(a, b):
+    st.write("Cache miss: expensive_computation(", a, ",", b, ") ran")
+    time.sleep(2)  # This makes the function take 2s to run
+    return a * b
+
+a = 2
+b = st.slider("Pick a number", 0, 10)  # 👈 Change this
+res = expensive_computation(a, b)
+
+st.write("Result:", res)
+```
+
+What you'll see:  
+
+* If you move the slider to a number Streamlit hasn't seen before, you'll have a cache miss again. And every subsequent rerun with the same number will be a cache hit, of course.
+* If you move the slider back to a number Streamlit has seen before, the cache is hit and the app is fast as expected.
+
+In computer science terms, what is happening here is that [`@st.cache`](api.html#streamlit.cache) is [memoizing](https://en.wikipedia.org/wiki/Memoization) `expensive_computation(a, b)`.
+
+But now let's go one step further! Try the following:
+
+1. Move the slider to a number you haven't tried before, such as 9.
+2. Pretend you're another user by opening another browser tab pointing to your Streamlit app (usually at http://localhost:8501)
+3. In the new tab, move the slider to 9.
+
+Notice how this is actually a cache hit! That is, you don't actually see the "Cache miss" text on the second tab even though that second user never moved the slider to 9 at any point prior to this.
+
+This happens because the Streamlit cache is global to all users. So everyone contributes to everyone else's performance.
+
+## Example 6: Mutating cached values
+
+As mentioned in the How Caching Works section, the Streamlit cache stores items by reference. This allows the Streamlit cache to support structures that aren't memory-managed by Python, such as TensorFlow objects. However, it can also lead to unexpected behavior — which is why Streamlit has a few checks to guide developers in the right direction. Let's look into those checks now.
```

Comment:
"How Caching Works" should be link

Reply:
Yea, this is a miss, but I have no idea where this is supposed to link? If this needs to link to the Advanced Caching topic, then it needs to go in the other PR -- not in this one. Since we split the PRs to try and rush this out the door last week. We can re-combine the PRs to simplify.